docs/architecture/adr-009-modular-antehandler.md

```diff
@@ -0,0 +1,261 @@
+
+#### Default Workflow:
+
+##### SDK code:
+
+```golang
+func Chainer(order []AnteHandler) AnteHandler {
+    return func(ctx Context, tx Tx, simulate bool) (newCtx Context, err error) {
+        for _, ante := range order {
+            ctx, err := ante(ctx, tx, simulate)
+            if err != nil {
+                return ctx, err
+            }
+        }
+        return ctx, err
+    }
+}
+```
+
+```golang
+func VerifySignatures(ctx Context, tx Tx, simulate bool) (newCtx Context, err error) {
+    // verify signatures
+    // Returns InvalidSignature Result and abort=true if sigs invalid
+    // Return OK result and abort=false if sigs are valid
+}
+
+func ValidateMemo(ctx Context, tx Tx, simulate bool) (newCtx Context, err error) {
+    // validate memo
+}
+
+AuthModuleAnteHandler := Chainer([]AnteHandler{VerifySignatures, ValidateMemo})
+```
+
+```golang
+func DeductFees(ctx Context, tx Tx, simulate bool) (newCtx Context, err error) {
+    // Deduct fees from tx
+    // Abort if insufficient funds in account to pay for fees
+}
+
+func CheckMempoolFees(ctx Context, tx Tx, simulate bool) (newCtx Context, err error) {
+    // If CheckTx: Abort if the fees are less than the mempool's minFee parameter
+}
+
+DistrModuleAnteHandler := Chainer([]AnteHandler{CheckMempoolFees, DeductFees})
+```
+
+```golang
+type ModuleManager struct {
+    // other fields
+    AnteHandlerOrder []AnteHandler
+}
+
+func (mm ModuleManager) GetAnteHandler() AnteHandler {
+    retun Chainer(mm.AnteHandlerOrder)
+}
+```
+
+##### User Code:
+
+```golang
+moduleManager.SetAnteHandlerOrder([]AnteHandler(AuthModuleAnteHandler, DistrModuleAnteHandler))
+
+app.SetAnteHandler(mm.GetAnteHandler())
+```
+
+#### Custom Workflow
+
+##### User Code
+
+```golang
+func CustomSigVerify(ctx Context, tx Tx, simulate bool) (newCtx Context, err error) {
+    // do some custom signature verification logic
+}
+```
+
+```golang
+// Micro-functions allow users to change order of when they get executed, and swap out default ante-functionality with their own custom logic.
+// Note that users can still chain the default distribution module handler, and auth micro-function along with their custom ante function
+moduleManager.SetAnteHandlerOrder([]AnteHandler(ValidateMemo, CustomSigVerify, DistrModuleAnteHandler))
+```
+
+Pros:
+1. Allows for ante functionality to be as modular as possible.
+2. For users that do not need custom ante-functionality, there is little difference between how antehandlers work and how BeginBlock and EndBlock work in ModuleManager.
+3. Still easy to understand
+
+Cons:
+1. Cannot wrap antehandlers with decorators like you can with Weave.
+
+### Simple Decorators
+
+This approach takes inspiration from Weave's decorator design while trying to minimize the number of breaking changes to the SDK and maximizing simplicity.
```

Comment:
can you explain here in the intro, what's the difference with what's under the `Decorator Pattern` section? 

```
changes to the SDK and maximizing simplicity. The difference(s) is/are ...
```